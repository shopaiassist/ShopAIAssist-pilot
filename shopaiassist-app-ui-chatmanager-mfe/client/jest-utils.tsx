import React, { JSXElementConstructor, ReactElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouterProps } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { I18nextProvider } from 'react-i18next';
import { merge } from 'ts-deepmerge';
import { configureStore, EnhancedStore } from '@reduxjs/toolkit';
import { ByRoleMatcher, render, RenderResult } from '@testing-library/react';
import { reducer, RootState, store as baseStore } from './src/store/store';
import i18nInstance from './src/utils/i18n';

type InitialEntries = MemoryRouterProps['initialEntries'];
type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P];
};
type PartialRootState = DeepPartial<RootState>;

interface RenderOptions {
  entries?: InitialEntries;
  state?: PartialRootState;
  store?: EnhancedStore;
}

interface SnapshotOptions {
  isModal?: boolean | ByRoleMatcher;
  state?: PartialRootState;
}

interface WrapperProps {
  children: ReactElement;
}

// export const mockApi = (apiInstance: AxiosInstance) => ({
//   delete: jest.spyOn(apiInstance, 'delete'),
//   get: jest.spyOn(apiInstance, 'get'),
//   post: jest.spyOn(apiInstance, 'post'),
//   put: jest.spyOn(apiInstance, 'put'),
//   patch: jest.spyOn(apiInstance, 'patch'),
// });

/** Redux store with optional default content */
export const mockStore = (state: PartialRootState = {}): EnhancedStore =>
  configureStore({ preloadedState: merge(baseStore.getState(), state), reducer: reducer as never });

/** Render function with access to override store/state */
export const renderAll = (ui: ReactElement, { entries, state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const Wrapper = ({ children }: WrapperProps) => (
    <MemoryRouter initialEntries={entries}>
      <Provider store={data}>
        <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
      </Provider>
    </MemoryRouter>
  );
  return render(ui, { wrapper: Wrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with i18n provider */
export const renderWithI18n = (ui: ReactElement): RenderResult => {
  const I18nWrapper = ({ children }: WrapperProps) => <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>;
  return render(ui, { wrapper: I18nWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Render function with access to override store/state */
export const renderWithRedux = (ui: ReactElement, { state = {}, store }: RenderOptions = {}): RenderResult => {
  const data = store || mockStore(state);
  const ReduxWrapper = ({ children }: WrapperProps) => (
    <Provider store={data}>
      <I18nextProvider i18n={i18nInstance}>{children}</I18nextProvider>
    </Provider>
  );
  return render(ui, { wrapper: ReduxWrapper as JSXElementConstructor<{ children: React.ReactNode }> });
};

/** Get content to test against a snapshot */
export const getSnapshot = (ui: ReactElement, state?: PartialRootState): ChildNode =>
  renderAll(ui, { state }).container.firstChild as ChildNode;

/** Get modal content to test against a snapshot */
export const getSnapshotModal = (ui: ReactElement, state?: PartialRootState, role?: ByRoleMatcher): ChildNode =>
  renderAll(ui, { state }).getByRole(role || 'presentation');

/** Helper function to generate snapshots */
export const snapshot = (ui: ReactElement, { isModal, state }: SnapshotOptions = {}) => {
  const role = typeof isModal === 'boolean' ? undefined : isModal;
  // eslint-disable-next-line no-extra-boolean-cast
  const snapshot = !!isModal ? getSnapshotModal(ui, state, role) : getSnapshot(ui, state);
  expect(snapshot).toMatchSnapshot();
};

// re-export everything
export * from '@testing-library/react';
