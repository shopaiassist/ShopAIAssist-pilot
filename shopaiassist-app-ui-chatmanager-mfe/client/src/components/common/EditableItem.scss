@import "../../utils/styles/imports";

.editable-item {
   display: flex;
   flex-direction: row;
   gap: 1rem;

   .button-container {
      display: flex;
      flex-direction: row;
      gap: .25rem;
      .editable-item-button::part(control) {
         padding: .5rem
      }
   }
}



