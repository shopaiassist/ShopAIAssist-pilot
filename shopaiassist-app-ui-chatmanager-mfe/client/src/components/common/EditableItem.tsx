import { Button, Icon, Textfield } from '@on/core-components/react';
import useEditableItem from '../../hooks/useEditableItem';

import './EditableItem.scss';
interface SelectableItemProps {
  initialValue: string;
  onAccept: (newName: string) => void;
  onCancel?: () => void;
}

/**
 * A component that provides an editable text field with accept and cancel buttons.
 * Utilizes useEditableItem for handling name updates and submissions.
 *
 * @param {string} initialValue - The initial value for the editable text field.
 * @param {(newName: string) => void} onAccept - Callback function that is called when the new name is submitted.
 * @param {() => void} [onCancel] - Optional callback function that is called when editing is canceled.
 * @returns {JSX.Element} The rendered component with a text field and control buttons.
 */
const EditableItem = ({ initialValue, onAccept, onCancel }: SelectableItemProps) => {
  const [updateText, submitNewText] = useEditableItem(initialValue, onAccept);
  return (
    <div className="editable-item">
      <Textfield
        data-testId="editable-text-field"
        initialValue={initialValue}
        onChange={(event) => {
          updateText(event.target.value);
        }}
        autoFocus={true}
      ></Textfield>
      <div className="button-container">
        <Button className="editable-item-button" onClick={submitNewText} data-testId="editable-item-submit">
          <Icon iconName="check" />
        </Button>
        <Button
          className="editable-item-button"
          appearance="secondary"
          onClick={onCancel}
          data-testId="editable-item-cancel"
        >
          <Icon iconName="close" />
        </Button>
      </div>
    </div>
  );
};

export default EditableItem;
