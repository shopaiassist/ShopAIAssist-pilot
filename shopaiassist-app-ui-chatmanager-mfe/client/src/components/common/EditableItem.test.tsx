import { act, fireEvent, renderWithI18n, snapshot, waitFor } from '../../../jest-utils';
import EditableItem from './EditableItem';

const getMockedComponent = (opts: { initialValue: string; onAccept: (newName: string) => void }) => {
  return <EditableItem initialValue={opts.initialValue} onAccept={opts.onAccept} />;
};

describe('EditableItem Tests', () => {
  describe('render', () => {
    it('should render', async () => {
      snapshot(getMockedComponent({ initialValue: 'Hello', onAccept: () => {} }));
    });
  });

  describe('user action', () => {
    const mockOnAccept = jest.fn();
    it('should change name of item', async () => {
      const component = getMockedComponent({ initialValue: 'Hello', onAccept: mockOnAccept });
      const wrapper = renderWithI18n(component);

      const textField = wrapper.getByTestId(`editable-text-field`);
      textField.click();
      fireEvent.change(textField, { target: { value: 'new name' } });
      await act(() => waitFor(() => wrapper.rerender(component)));
      const submitButton = wrapper.getByTestId(`editable-item-submit`);
      submitButton.click();
      await act(() => waitFor(() => wrapper.rerender(component)));

      expect(mockOnAccept).toHaveBeenCalledWith('new name');
    });
  });
});
