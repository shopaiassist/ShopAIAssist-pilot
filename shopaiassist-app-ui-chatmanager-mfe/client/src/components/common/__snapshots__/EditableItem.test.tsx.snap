// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableItem Tests render should render 1`] = `
<div
  class="editable-item"
>
  <textarea
    data-testid="editable-text-field"
    initialvalue="Hello"
  />
  <div
    class="button-container"
  >
    <button
      class="editable-item-button"
      data-testid="editable-item-submit"
    >
      <div
        iconname="check"
      />
    </button>
    <button
      appearance="secondary"
      class="editable-item-button"
      data-testid="editable-item-cancel"
    >
      <div
        iconname="close"
      />
    </button>
  </div>
</div>
`;
